File: Binary_Tree_Postorder_Traversal.cpp
/**
 * Definition of TreeNode:
 * class TreeNode {
 * public:
 *     int val;
 *     TreeNode *left, *right;
 *     TreeNode(int val) {
 *         this->val = val;
 *         this->left = this->right = NULL;
 *     }
 * }
 */
class Solution {
    /**
     * @param root: The root of binary tree.
     * @return: Postorder in vector which contains node values.
     */
public:
    vector<int> postorderTraversal(TreeNode *root) {
        // write your code here
    //     vector<int> result;
    //     doOrder(root, result);
    //     return result;
    // }
    // void doOrder(TreeNode *node, vector<int>& vct){
    //     if(!node) return;
    //     if(node->left){
    //         doOrder(node->left, vct);
    //     }
    //     if(node->right){
    //         doOrder(node->right, vct);
    //     }
    //     vct.emplace_back(node->val);
    
        // write your code here using space(O(n)), time(O(n))
        // if(!root) return {};
        // stack<TreeNode*> path_stack;
        // path_stack.emplace(root);
        // TreeNode* pre = nullptr;
        // vector<int> result;
        // while(!path_stack.empty()){
        //     TreeNode* cur = path_stack.top();
        //     if(pre == nullptr || pre->left == cur || pre->right == cur){
        //         if(cur->left){
        //             path_stack.emplace(cur->left);
        //         }else if(cur->right){
        //             path_stack.emplace(cur->right);
        //         }else{
        //             result.emplace_back(cur->val);
        //             path_stack.pop();
        //         }
        //     }else if(cur->left == pre){
        //         if(cur->right){
        //             path_stack.emplace(cur->right);
        //         }else{
        //             result.emplace_back(cur->val);
        //             path_stack.pop();
        //         }
        //     }else{
        //         result.emplace_back(cur->val);
        //         path_stack.pop();
        //     }
        //     pre = cur;
        // }
        // return result;
        // write your code here using space(O(1)), time(O(n))
        vector<int> result;
        TreeNode dummy(INT_MIN);
        dummy.left = root;
        TreeNode* cur = &dummy;
        while(cur){
            if(!cur->left){
                cur = cur->right;
            }else{
                TreeNode* node = cur->left;
                while(node->right && node->right != cur){
                    node = node->right;
                }
                if(!node->right){
                    node->right = cur;
                    cur = cur->left;
                }else{
                    vector<int> res = dotravel(cur->left, node);
                    result.insert(result.end(), res.begin(), res.end());
                    node->right = nullptr;
                    cur = cur->right;
                }
            }
        }
        return result;
    }
    vector<int> dotravel(TreeNode* from, TreeNode* to){
        vector<int> res;
        TreeNode* cur = from;
        while(cur != to){
            res.emplace_back(cur->val);
            cur = cur->right;
        }
        res.emplace_back(cur->val);
        reverse(res.begin(), res.end());
        return res;
    }
};

File: Binary Tree Postorder Traversal.py
"""
Definition of TreeNode:
class TreeNode:
    def __init__(self, val):
        self.val = val
        self.left, self.right = None, None
"""


class Solution:
    """
    @param root: The root of binary tree.
    @return: Postorder in ArrayList which contains node values.
    """
    def postorderTraversal(self, root):
        # write your code here
        
        # def doTra(node, ret):
        #     if not node:
        #         return
        #     else:
        #         doTra(node.left, ret)
        #         doTra(node.right, ret)
        #         ret.append(node.val)
        
        # ret = []
        # doTra(root, ret)
        # return ret
        
        # ret = []
        # if not root:
        #     return ret
        
        # stack = []
        # node = root
        # while node:
        #     stack.append([node, 0])
        #     node = node.left
        # while stack:
        #     p = stack[-1]
        #     if p[1] == 0:
        #         ##error p[1] "==" 1
        #         p[1] = 1
        #         node = p[0].right
        #         while node:
        #             stack.append([node, 0])
        #             node = node.left
        #     else:
        #         ret.append(p[0].val)
        #         stack.pop()
        # return ret
        if not root:
            return []
        stack, pre = [root], None
        ret = []
        while stack:
            cur = stack[-1]
            if not pre or pre.left == cur or pre.right == cur:
                if cur.left:
                    stack.append(cur.left)
                elif cur.right:
                    stack.append(cur.right)
            elif cur.left == pre:
                if cur.right:
                    stack.append(cur.right)
            else:
                ret.append(cur.val)
                stack.pop()
            pre = cur
        return ret